tests/ci/ci.py

```diff
@@ -0,0 +1,637 @@
+            "stdout": process.stdout,
+            "stderr": process.stderr,
+            "return_code": process.returncode,
+        }
+
+    result: Dict[str, str] = {}
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        futures = [
+            executor.submit(run_docker_command, image, tag, arch)
+            for image, tag in image_name_tag.items()
+        ]
+
+        responses = [
+            future.result() for future in concurrent.futures.as_completed(futures)
+        ]
+        for resp in responses:
+            name, stdout, stderr, digest, arch = (
+                resp["image"],
+                resp["stdout"],
+                resp["stderr"],
+                resp["image_digest"],
+                resp["arch"],
+            )
+            if stderr:
+                if stderr.startswith("no such manifest"):
+                    result[name] = digest
+                else:
+                    print(f"Eror: Unknown error: {stderr}, {name}, {arch}")
+            elif stdout:
+                if "mediaType" in stdout:
+                    pass
+                else:
+                    print(f"Eror: Unknown response: {stdout}")
+                    assert False, "FIXME"
+            else:
+                print(f"Eror: No response for {name}, {digest}, {arch}")
+                assert False, "FIXME"
+    return result
+
+
+def check_and_update_for_early_style_check(run_config: dict) -> None:
+    """
+    This is temporary hack to start style check before docker build if possible
+    FIXME: need better solution to do style check as soon as possible and as fast as possible w/o dependency on docker job
+    """
+    jobs_to_do = run_config.get("jobs_data", {}).get("jobs_to_do", [])
+    docker_to_build = run_config.get("docker_data", {}).get("missing_multi", [])
+    if (
+        "Style check" in jobs_to_do
+        and docker_to_build
+        and "clickhouse/style-test" not in docker_to_build
+    ):
+        index = jobs_to_do.index("Style check")
+        jobs_to_do[index] = "Style check early"
+
+
+def main() -> int:
+    exit_code = 0
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+    args = parse_args(parser)
+
+    if (args.mark_success or args.pre or args.post) and not args.infile:
+        print("ERROR: need option --infile to be provided")
+        parser.print_help()
+        parser.exit(1)
+    if args.mark_success or args.pre or args.post or args.run:
+        assert args.job_name, "Job name must be provided via --job-name"
+
+    indata: Optional[Dict[str, Any]] = None
+    if args.infile:
+        indata = (
+            json.loads(args.infile)
+            if not os.path.isfile(args.infile)
+            else json.load(open(args.infile))
+        )
+        assert indata and isinstance(indata, dict), "Invalid --infile json"
+
+    result: Dict[str, Any] = {}
+
+    s3 = S3Helper()
+
+    if args.configure:
```

Comment:
it looks like a separate function to me. And every `elif` below too

Reply:
here's example of `subparser` for args https://github.com/ClickHouse/ClickHouse/pull/57203/files#diff-d3d6879ef42c74415ca77dabb8fb1ae78ab4cf8c5ea5b229dcdff0b821cc9c5f

Reply:
I added a todo about subparser. It looks like a good match for ci.py script.